// aws-crt/dist/index.js
"use strict";
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CrtError = exports.resource_safety = exports.promise = exports.platform = exports.mqtt5 = exports.mqtt = exports.iot = exports.io = exports.http = exports.eventstream = exports.crt = exports.crypto = exports.checksums = exports.cancel = exports.auth = void 0;
// This is the entry point for the AWS CRT nodejs native libraries
/* common libs */
const cancel = __importStar(require("./common/cancel"));
exports.cancel = cancel;
const platform = __importStar(require("./common/platform"));
exports.platform = platform;
const promise = __importStar(require("./common/promise"));
exports.promise = promise;
const resource_safety = __importStar(require("./common/resource_safety"));
exports.resource_safety = resource_safety;
/* node specific libs */
const auth = __importStar(require("./native/auth"));
exports.auth = auth;
const checksums = __importStar(require("./native/checksums"));
exports.checksums = checksums;
const crt = __importStar(require("./native/crt"));
exports.crt = crt;
const crypto = __importStar(require("./native/crypto"));
exports.crypto = crypto;
const eventstream = __importStar(require("./native/eventstream"));
exports.eventstream = eventstream;
const http = __importStar(require("./native/http"));
exports.http = http;
const io = __importStar(require("./native/io"));
exports.io = io;
const iot = __importStar(require("./native/iot"));
exports.iot = iot;
const mqtt = __importStar(require("./native/mqtt"));
exports.mqtt = mqtt;
const mqtt5 = __importStar(require("./native/mqtt5"));
exports.mqtt5 = mqtt5;
const error_1 = require("./native/error");
Object.defineProperty(exports, "CrtError", { enumerable: true, get: function () { return error_1.CrtError; } });
